mypy/plugin.py

```diff
@@ -1,40 +1,121 @@
+analyzer will be removed soon). In particular, plugins should not hold global
+state, and should always call add_plugin_dependency() in plugin hooks called
+during semantic analysis. See the method docstring for more details.
+
+There is no dedicated cache storage for plugins, but plugins can store
+per-TypeInfo data in a special .metadata attribute that is serialized to the
+mypy caches between incremental runs. To avoid collisions between plugins, they
+are encouraged to store their state under a dedicated key coinciding with
+plugin name in the metadata dictionary. Every value stored there must be
+JSON-serializable.
+
+## New semantic analyzer
+
+The new semantic analyzer (enabled through the --new-semantic-analyzer flag)
+changes how plugins are expected to work in several notable ways:
+
+1. The order of processing modules is different. The old semantic analyzer
+   processed modules in textual order, one module at a time. The new semantic
+   analyzer first processes the module top levels, including bodies of any
+   top-level classes. Functions and methods are processed only after module
+   top levels have been finished. If there is an import cycle, all module
+   top levels in the cycle are processed before processing any functions or
+   methods. Each unit of processing (a module top level or a function/method)
+   is called a *target*.
+
+   This also means that function signatures in the same module have not been
+   analyzed yet when analyzing the module top level. If you need access to
+   a function signature, you'll need to explicitly analyze the signature first
+   using `type_anal()`.
+
+2. Each target can be processed multiple times. This may happen if some forward
+   references are not ready yet, for example. This means that semantic analyzer
+   related plugin hooks can be called multiple times for the same full name.
+   These plugin methods must thus be idempotent.
+
+3. The `anal_type` API function returns None if some part of the type is not
+   available yet. If this happens, the current target being analyzed will be
+   *deferred*, which means that it will be processed again soon, in the hope
+   that additional dependencies will be available. This may happen if there are
+   forward references to types or inter-module references to types within an
+   import cycle.
+
+   Note that if there is a circular definition, mypy may decide to stop
+   processing to avoid an infinite number of iterations. When this happens,
+   `anal_type` will generate an error and return an `AnyType` type object
+   during the final iteration (instead of None).
+
+4. There is a new API method `defer()`. This can be used to explicitly request
+   the current target to be reprocessed one more time. You don't need this
+   to call this if `anal_type` returns None, however.
+
+5. There is a new API property `final_iteration`, which is true once the
+   maximum semantic analysis iteration count has been reached. You must never
```

Comment:
Maybe mention this can be also `True` sooner if mypy detected no progress during last iteration?